File: c/ch4/ecjac.c
static char help[] = "Newton's method for a two-variable system.  Implements analytical Jacobian and a struct to hold a parameter.\n";

#include <petsc.h>

typedef struct {
  PetscReal  b;
} AppCtx;

extern PetscErrorCode FormFunction(SNES, Vec, Vec, void*);
extern PetscErrorCode FormJacobian(SNES, Vec, Mat, Mat, void*);

//STARTMAIN
int main(int argc,char **argv) {
  SNES   snes;         // nonlinear solver
  Vec    x,r;          // solution, residual vectors
  Mat    J;
  AppCtx user;

  PetscCall(PetscInitialize(&argc,&argv,NULL,help));
  user.b = 2.0;

  PetscCall(VecCreate(PETSC_COMM_WORLD,&x));
  PetscCall(VecSetSizes(x,PETSC_DECIDE,2));
  PetscCall(VecSetFromOptions(x));
  PetscCall(VecDuplicate(x,&r));

  PetscCall(MatCreate(PETSC_COMM_WORLD,&J));
  PetscCall(MatSetSizes(J,PETSC_DECIDE,PETSC_DECIDE,2,2));
  PetscCall(MatSetFromOptions(J));
  PetscCall(MatSetUp(J));

  PetscCall(SNESCreate(PETSC_COMM_WORLD,&snes));
  PetscCall(SNESSetFunction(snes,r,FormFunction,&user));
  PetscCall(SNESSetJacobian(snes,J,J,FormJacobian,&user));
  PetscCall(SNESSetFromOptions(snes));

  PetscCall(VecSet(x,1.0));            // initial iterate
  PetscCall(SNESSolve(snes,NULL,x));
  PetscCall(VecView(x,PETSC_VIEWER_STDOUT_WORLD));

  PetscCall(SNESDestroy(&snes));
  PetscCall(MatDestroy(&J));
  PetscCall(VecDestroy(&x));
  PetscCall(VecDestroy(&r));
  PetscCall(PetscFinalize());
  return 0;
}
//ENDMAIN

//STARTFORM
PetscErrorCode FormFunction(SNES snes, Vec x, Vec F, void *ctx) {
    AppCtx          *user = (AppCtx*)ctx;
    const PetscReal b = user->b, *ax;
    PetscReal       *aF;

    PetscCall(VecGetArrayRead(x,&ax));
    PetscCall(VecGetArray(F,&aF));
    aF[0] = (1.0 / b) * PetscExpReal(b * ax[0]) - ax[1];
    aF[1] = ax[0] * ax[0] + ax[1] * ax[1] - 1.0;
    PetscCall(VecRestoreArrayRead(x,&ax));
    PetscCall(VecRestoreArray(F,&aF));
    return 0;
}

PetscErrorCode FormJacobian(SNES snes, Vec x, Mat J, Mat P, void *ctx) {
    AppCtx           *user = (AppCtx*)ctx;
    const PetscReal  b = user->b, *ax;
    PetscReal        v[4];
    PetscInt         row[2] = {0,1}, col[2] = {0,1};

    PetscCall(VecGetArrayRead(x,&ax));
    v[0] = PetscExpReal(b * ax[0]);  v[1] = -1.0;
    v[2] = 2.0 * ax[0];              v[3] = 2.0 * ax[1];
    PetscCall(VecRestoreArrayRead(x,&ax));
    PetscCall(MatSetValues(P,2,row,2,col,v,INSERT_VALUES));
    PetscCall(MatAssemblyBegin(P,MAT_FINAL_ASSEMBLY));
    PetscCall(MatAssemblyEnd(P,MAT_FINAL_ASSEMBLY));
    if (J != P) {
        PetscCall(MatAssemblyBegin(J,MAT_FINAL_ASSEMBLY));
        PetscCall(MatAssemblyEnd(J,MAT_FINAL_ASSEMBLY));
    }
    return 0;
}
//ENDFORM

File: python/ch4/ecjac.py
#!/usr/bin/env python3
'''Newton's method for a two-variable system.  Implements analytical
Jacobian and a struct to hold a parameter.'''

import sys
from numpy import exp
import petsc4py
petsc4py.init(sys.argv)  # must come before "import PETSc"
from petsc4py import PETSc

class AppCtx:
    '''Holds parameter b.  Supplies call-back functions.'''

    def __init__(self, b=1.0):
        self.b = b

    def formFunction(self, _snes, X, F):
        '''Evaluate F(X).'''
        ax = X.getArray(readonly=1)
        af = F.getArray(readonly=0)
        af[0] = (1.0 / self.b) * exp(self.b * ax[0]) - ax[1]
        af[1] = ax[0] * ax[0] + ax[1] * ax[1] - 1.0

    def formJacobian(self, _snes, X, Jmat, Pmat):
        '''Evaluate J(X) where J = F'.'''
        ax = X.getArray(readonly=1)
        row = [0,1]
        col = row
        v = [exp(self.b * ax[0]),  -1.0,
             2.0 * ax[0],          2.0 * ax[1]]
        Pmat.setValues(row,col,v)
        Pmat.assemble()
        if Jmat != Pmat:
            Jmat.assemble()

ctx = AppCtx(b=2.0)

x = PETSc.Vec()
x.create(PETSc.COMM_WORLD)
x.setSizes(2)
x.setFromOptions()
x.set(1.0)
r = x.duplicate()

J = PETSc.Mat()
J.create(PETSc.COMM_WORLD)
J.setSizes((2,2))
J.setFromOptions()
J.setUp()

snes = PETSc.SNES()
snes.create(PETSc.COMM_WORLD)
snes.setFunction(ctx.formFunction, r)
snes.setJacobian(ctx.formJacobian, J)
snes.setFromOptions()
snes.solve(None,x)
x.view()
